src/storage/ob_partition_parallel_merge_ctx.cpp

```diff
@@ -160,6 +162,473 @@ int ObParallelMergeCtx::init_serial_merge()
   return ret;
 }
 
+// to check whether the ranges are contiguous (valid)
+static bool contiguous_ranges(const common::ObIArray<common::ObStoreRange>& ranges)
+{
+  bool bret = true;
+
+  if (OB_UNLIKELY(ranges.count()) < 1) {
+    STORAGE_LOG(WARN, "Unexpected bad ranges (empty).", K(ranges));
+    bret = false;
+  } else if (OB_UNLIKELY(!ranges.at(0).get_start_key().is_min())) {
+    STORAGE_LOG(WARN, "Unexpected bad ranges (first start key not min).", K(ranges));
+    bret = false;
+  } else if (OB_UNLIKELY(!ranges.at(ranges.count() - 1).get_end_key().is_max())) {
+    STORAGE_LOG(WARN, "Unexpected bad ranges (last end key not max).", K(ranges));
+    bret = false;
+  } else {
+    for (int64_t i = 1; i < ranges.count(); i++) {
+      if (OB_UNLIKELY(ranges.at(i).get_start_key() != ranges.at(i - 1).get_end_key())) {
+        STORAGE_LOG(WARN, "Unexpected bad ranges (not contiguous).", K(ranges));
+        bret = false;
+        break;
+      }
+    }
+  }
+
+  return bret;
+}
+
+int ObParallelMergeCtx::get_ranges_by_base_sstable(
+    const int64_t tablet_size, common::ObIArray<common::ObStoreRange>& ranges)
+{
+  int ret = OB_SUCCESS;
+
+  int64_t row_cnt = first_sstable_->get_total_row_count();
+  const common::ObIArray<blocksstable::MacroBlockId>& macro_block_ids = first_sstable_->get_macro_block_ids();
+  const int64_t macro_block_cnt = macro_block_ids.count();
+  if (row_cnt <= 0) {
+    common::ObStoreRange range;
+    range.set_whole_range();
+    if (OB_FAIL(ranges.push_back(range))) {
+      STORAGE_LOG(WARN, "Failed to push range to ranges.", K(ret));
+    }
+    STORAGE_LOG(INFO, "Empty base sstable.", K(ret), K(row_cnt));
+  } else {
+    if (OB_UNLIKELY(macro_block_cnt <= 0)) {
+      ret = OB_ERR_UNEXPECTED;
+      STORAGE_LOG(WARN, "Unexpected empty base sstable.", K(ret), K(row_cnt), K(macro_block_cnt));
+    } else {
+      common::ObStoreRowkey rowkey;
+      common::ObStoreRange range;
+      range.get_end_key().set_min();
+      range.set_left_open();
+      range.set_right_closed();
+      blocksstable::ObFullMacroBlockMeta blk_meta;
+      int64_t num_blocks_per_range = tablet_size / OB_FILE_SYSTEM.get_macro_block_size();
```

Comment:
the tablet_size should be checked as a input parameters, since we cannot deal with the tablet_size which smaller than macro block size